tests/std/tests/P2321R2_views_zip/test.cpp

```diff
@@ -0,0 +1,838 @@
+                        assert(diff2 == static_cast<difference_type>(TestContainerType::smallest_array_size));
+                    } else if constexpr (std::constructible_from<difference_type,
+                                             decltype(TestContainerType::smallest_array_size)>) {
+                        assert(diff1 == -difference_type{TestContainerType::smallest_array_size});
+                        assert(diff2 == difference_type{TestContainerType::smallest_array_size});
+                    }
+                }
+            }
+        };
+
+        // Validate iterators and sentinels
+        if constexpr (CanMemberBegin<ZipType> && CanMemberEnd<ZipType>) {
+            validate_iterators_lambda.template operator()<decltype(tuple_element_arr), ZipType, AllView<RangeTypes>...>(
+                zipped_range, tuple_element_arr);
+        }
+
+        if constexpr (CanMemberBegin<const ZipType> && CanMemberEnd<const ZipType>) {
+            validate_iterators_lambda
+                .template operator()<decltype(const_tuple_element_arr), const ZipType, const AllView<RangeTypes>...>(
+                    std::as_const(zipped_range), const_tuple_element_arr);
+        }
+    }
+
+    return true;
+}
+
+// zip_view<RangeTypes...> is sensitive to the following:
+//
+//   - The categories of RangeTypes...
+//   - The commonality of RangeTypes...
+//   - Whether or not the size() member exists for each of RangeTypes..., as it is used by std::ranges::size()
+//   - The sizeof...(RangeTypes)
+//
+// Other conditions are irrelevant.
+
+template <class Category, class Element, test::Sized IsSized, test::Common IsCommon>
+using test_range = test::range<Category, Element, IsSized,
+    test::CanDifference{std::derived_from<Category, std::random_access_iterator_tag>}, IsCommon,
+    (to_bool(IsCommon) ? test::CanCompare::yes
+                       : test::CanCompare{std::derived_from<Category, std::forward_iterator_tag>})>;
+
+constexpr std::array default_int_array{0, 1, 2, 3, 4, 5, 6};
+
+constexpr single_element_test_container<int, 7> single_element_container_instance{default_int_array};
+constexpr three_element_test_container<int, 7, float, 9, char, 5> three_element_container_instance{default_int_array,
+    std::array{0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f}, std::array{'a', 'b', 'c', 'd', 'e'}};
+
+// The typical instantiator struct isn't templated like this, so that it could then be used with the
+// existing testing machinery defined in range_algorithm_support.hpp. We can't use that stuff here,
+// however, because std::views::zip takes multiple template parameters.
+//
+// The idea here might be useful for other view types which take multiple range parameters. Feel
+// free to move something similar into range_algorithm_support.hpp if you feel that this could be
+// used in another test suite.
+
+template <bool IsMoveOnly>
+struct range_type_solver {

```

Comment:
These ain't structures
```suggestion
class range_type_solver {
```